packages/@sanity/form-builder/src/inputs/ReferenceInput/usePreviewSnapshot.ts

```diff
@@ -0,0 +1,48 @@
+import {useEffect, useState} from 'react'
+import {Reference} from '@sanity/types'
+import {tap} from 'rxjs/operators'
+import {Observable} from 'rxjs'
+
+type SnapshotState = {
+  isLoading: boolean
+  snapshot: null | PreviewSnapshot
+}
+
+const LOADING_SNAPSHOT: SnapshotState = {
+  isLoading: true,
+  snapshot: null,
+}
+
+const NULL_SNAPSHOT: SnapshotState = {
+  isLoading: false,
+  snapshot: null,
+}
+
+type PreviewSnapshot = {
+  _id: string
+  _type: string
+  title: string
+  description: string
+}
+
+export function usePreviewSnapshot(
```

Comment:
Since this function takes a `Reference` to provide the preview data, should it be named `useReferencePreviewSnapshot`?

Reply:
Agree. But not part of this PR (I think this branch was outdated when you reviewed this)